reposerver/repository/repository.go

```diff
@@ -754,8 +755,103 @@ func (s *Service) runManifestGenAsync(ctx context.Context, repoRoot, commitSHA,
+					return
+				}
+				// Check for this limitation is taken from above
+				if refSourceMapping.Chart != "" {
+					ch.errCh <- fmt.Errorf("source has a 'chart' field defined, but Helm charts are not yet not supported for 'ref' sources")
+					return
+				}
+
+				// The ref is available, now - checking the tree out
+				normalizedRepoURL := git.NormalizeGitURL(refSourceMapping.Repo.Repo)
+				/* - @crenshaw-dev suggested to drop this limitation here
+				repoRef, ok := repoRefs[normalizedRepoURL]
+				if ok {
+					if repoRef.revision != refSourceMapping.TargetRevision {
+						ch.errCh <- fmt.Errorf("cannot reference multiple revisions for the same repository (%s references %q while %s references %q)", refVar, refSourceMapping.TargetRevision, repoRef.key, repoRef.revision)
+						return
+					}
+				}
+				*/
+
+				// Check whether the source ref in copyfrom is pointing at the current appsource
+				// Current AppSource's ref: q.Ref
+				if q.ApplicationSource.Ref == copySpec.Ref {
+					ch.errCh <- fmt.Errorf("from cannot reference the ApplicationSource it's declared at: %s", q.ApplicationSource.Ref)
+					return
+				}
+
+				gitClient, referencedCommitSHA, err := s.newClientResolveRevision(&refSourceMapping.Repo, refSourceMapping.TargetRevision)
+				if err != nil {
+					ch.errCh <- fmt.Errorf("failed to get git client for repo %s", refSourceMapping.Repo.Repo)
+					return
+				}
+
+				if git.NormalizeGitURL(q.ApplicationSource.RepoURL) == normalizedRepoURL && commitSHA != referencedCommitSHA {
+					ch.errCh <- fmt.Errorf("cannot reference a different revision of the same repository (%s references %q which resolves to %q while the application references %q which resolves to %q)", refVar, refSourceMapping.TargetRevision, referencedCommitSHA, q.Revision, commitSHA)
+					return
+				}
+				closer, err := s.repoLock.Lock(gitClient.Root(), referencedCommitSHA, true, func() (goio.Closer, error) {
+					return s.checkoutRevision(gitClient, referencedCommitSHA, s.initConstants.SubmoduleEnabled)
+				})
+				if err != nil {
+					log.Errorf("failed to acquire lock for referenced source %s", normalizedRepoURL)
+					ch.errCh <- err
+					return
+				}
+				defer func(closer goio.Closer) {
+					err := closer.Close()
+					if err != nil {
+						log.Errorf("Failed to release repo lock: %v", err)
+					}
+				}(closer)
+
+				// Collect the copy's input data
+				srcRepoRoot := gitClient.Root()
+				dstRepoRoot := repoRoot
+				log.WithFields(log.Fields{
+					"ref": copySpec.Ref,
+					"srcpath": copySpec.SourcePath,
+					"dstpath": copySpec.DestinationPath,
+					"hardfail": copySpec.Hardfail,
+					"srcRepoRoot": srcRepoRoot,
+					"dstRepoRoot": dstRepoRoot,
+				}).Info("Starting repo copy")
+
+				// do the copy
+				err = copyFilesystemSubtree(srcRepoRoot, copySpec.SourcePath, dstRepoRoot, copySpec.DestinationPath, copySpec.Hardfail)
```

Comment:
Can we release the lock at this point? If so, we can drop the `cannot reference a different revision of the same repository` check, because there's no risk of a deadlock - we're all done here before the main source's revision is checked out.

Reply:
@crenshaw-dev  I'm not very familiar with the higher level workflow, so I cannot be a judge of that. If you advise the lock is to be released, I can see to it.

Reply:
`ffd8627c` should be fixing this. For some reason I had the impression that the deferred closer at L824 does this job automagically. Sorry for the confusion.